Index: prelink/testsuite/functions.sh
===================================================================
--- prelink.orig/testsuite/functions.sh
+++ prelink/testsuite/functions.sh
@@ -1,7 +1,7 @@
 #!/bin/bash
-CC="${CC:-gcc} ${LINKOPTS}"
+CC="${CC:-gcc} ${LINKOPTS} -Wl,--copy-dt-needed-entries" ; echo "CC=$CC"
-CCLINK=${CCLINK:-${CC} -Wl,--dynamic-linker=`echo ./ld*.so.*[0-9]`}
+CCLINK="${CCLINK:-${CC} -Wl,--dynamic-linker=`echo ./ld*.so.*[0-9]`} -Wl,--copy-dt-needed-entries"
-CXX="${CXX:-g++} ${LINKOPTS}"
+CXX="${CXX:-g++} ${LINKOPTS} -Wl,--copy-dt-needed-entries" ; echo "CXX=$CXX"
-CXXLINK=${CXXLINK:-${CXX} -Wl,--dynamic-linker=`echo ./ld*.so.*[0-9]`}
+CXXLINK="${CXXLINK:-${CXX} -Wl,--dynamic-linker=`echo ./ld*.so.*[0-9]`} -Wl,--copy-dt-needed-entries"
 PRELINK=${PRELINK:-../src/prelink -c ./prelink.conf -C ./prelink.cache --ld-library-path=. --dynamic-linker=`echo ./ld*.so.*[0-9]`}
 srcdir=${srcdir:-`dirname $0`}
